ssl/bio_ssl.c

```diff
@@ -284,6 +284,7 @@ static long ssl_ctrl(BIO *b, int cmd, long num, void *ptr)
             ssl_free(b);
             if (!ssl_new(b))
                 return 0;
+            bs = BIO_get_data(b);
```

Comment:
@mattcaswell it seems like this bug was introduced way back in 2016 when the BIO was made opaque in commit a146ae55ba479a5c7aa2a6afba1b2b93102a152c and line 291 below [was changed as follows](https://github.com/openssl/openssl/commit/a146ae55ba479a5c7aa2a6afba1b2b93102a152c#diff-aa46282ee9d01cf3c7f3d1f9a4c3bfafR340):

```diff
@@ -330,17 +335,17 @@ static long ssl_ctrl(BIO *b, int cmd, long num, void *ptr)
...
-        ((BIO_SSL *)b->ptr)->ssl = ssl;
+        bs->ssl = ssl;
```

It's hard to believe that this bug went unnoticed for such a long time (or is this ctrl used only rarely?). This error should be fixed for master and 1.1.1 before the next release IMHO.

Reply:
I suppose it is usually (practically always given there were no bug reports before) called only once on a SSL BIO so this code does not come into play.